lib/text-class/src/Data/Text/Class.hs

```diff
@@ -86,3 +102,74 @@ instance ToText Natural where
 -- | Decode the specified text with a 'Maybe' result type.
 fromTextMaybe :: FromText a => Text -> Maybe a
 fromTextMaybe = either (const Nothing) Just . fromText
+
+-- | Represents a case style for multi-word strings.
+data CaseStyle
+    = CamelCase
+      -- ^ A string in the style of "doNotRepeatYourself"
+    | PascalCase
+      -- ^ A string in the style of "DoNotRepeatYourself"
+    | KebabLowerCase
+      -- ^ A string in the style of "do-not-repeat-yourself"
+    | SnakeLowerCase
+      -- ^ A string in the style of "do_not_repeat_yourself"
+    | SnakeUpperCase
+      -- ^ A string in the style of "DO_NOT_REPEAT_YOURSELF"
+    | SpacedLowerCase
+      -- ^ A string in the style of "do not repeat yourself"
+    deriving (Bounded, Enum, Eq, Generic, Show)
+
+-- | Converts the given value to text, according to the specified 'CaseStyle'.
+--
+-- This function guarantees to satisfy the following property:
+--
+-- > fromTextToBoundedEnum s (toTextFromBoundedEnum s a) == Right a
+--
+toTextFromBoundedEnum :: Show a => CaseStyle -> a -> Text
+toTextFromBoundedEnum cs = T.pack . toCaseStyle cs . Casing.fromHumps . show
```

Comment:
I wonder if there be some constraint added to ensure `a` is a sum type of nullary constructors? So that `show` does not produce multiple words.

Reply:
Some like `GSumType (Rep a) =>` using generics, where `GSumType` is to define ?

Reply:
> I wonder if there could be some constraint added to ensure `a` is a sum type of nullary constructors? So that `show` does not produce multiple words.

Good question.

Actually, when attempting to make a derived instance of `Enum`, the compiler will check that the type is an enumeration type (consisting of one or more nullary, non-GADT constructors).

If we attempt to do this:
```hs
data Wibble
    = Foo
    | Bar
    | Baz String
   deriving (Bounded, Enum, Eq, Generic, Ord, Show)
```
Then the compiler will complain with:
```
[Error] Can't make a derived instance of ‘Enum Wibble’:
‘Wibble’ must be an enumeration type.
(an enumeration consists of one or more nullary, non-GADT constructors).

So we could add the `Enum` and `Bounded` constraints to `toTextFromBoundedEnum`. However, we have to convince the compiler to allow these constraints (due to the `-fwarn-redundant-constraints`).